### PR TITLE
CRM-20936: Hide statuses on backoffice contribution form

### PR DESCRIPTION
**Pull request by monishdeb**
* [CRM-20936: Hide statuses on backoffice contribution form](https://issues.civicrm.org/jira/browse/CRM-20936)

### CODE REVIEW

**Comment by pradpnayak** [comment]
tested, works fine.
Before applying patch
beforepatch

After Patch
afterpatch

**Comment by eileenmcnaughton** [comment]
Having those pledge statuses in there & hack-handled is weird - we really should get this done!

https://issues.civicrm.org/jira/browse/CRM-20913

We probably should put a centralised constant for those 'New available for new' statuses & refer to that rather than defining on the form - although perhaps the case is that they really are form-layer logic & only apply to one form?

**Comment by monishdeb** [comment]
@eileenmcnaughton I have updated the PR. Can you please check now?

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit can you review this once @monishdeb  has reviewed #10737 - it should need minor changes once that is merged. In principle I think this is a good change & it centralises code in a good way

**Comment by monishdeb** [comment]
@eileenmcnaughton I have updated the PR after merging #10737 But I have a doubt if we need to delete ```In Progress``` and ```Overdue``` status from Contribution Status's option values too?

**Comment by pradpnayak** [comment]
@monishdeb you missed to make changes in Membership Renewal form.

**Comment by monishdeb** [comment]
Thanks @pradpnayak .. I have updated the PR.

File: CRM/Contribute/Form/Contribution.php
Comment (jitendrapurohit) [inline CRM/Contribute/Form/Contribution.php]: seems to be a problem here where we set `$component` as `pledge` only when the first if block is not `TRUE` which checks whether we are in update mode - `$this->_id`. So, I think `$component` will not be `pledge` even if we are updating any pledge payment? For eg: 

1. Record a payment for pledge with `pending` status.
2. Click `Record Payment` again for the above pending payment. Shouldn't `$component` be equal to `pledge` here?

**Comment by seamuslee001** [comment]
So i took a look at this, @monishdeb seems to be mostly ok from what i can tell, the only thing is looking at what was removed in CRM/Membership/Form/Membership.php would suggest we need to be unsetting certain status when we are recording a membership payment for a pending  online membership which don't seem to be doing in your consolidated function.  Also when doing an ordinary offline renewal i note that in the list of payment statues we have In Progress, Overdue, Partially Paid,  which doesn't seem to me to be accurate. Also as i'm not as versed in the whole Recording Payment could you just check what Jitendra has said on the PR re pledge payments.

**Comment by monishdeb** [comment]
Thanks @seamuslee001 @jitendrapurohit for your feedback. Will fix these remaining issues and will let you know

**Comment by monishdeb** [comment]
@seamuslee001 @jitendrapurohit I have made the fixes - https://github.com/civicrm/civicrm-core/pull/10718/commits/866a94d4f3d4b9d715b25b552cce5a9c46d19b2e

So for membership payment, I have excluded ```In Progress``` and  ```Overdue``` status, not partially paid because we will be supporting partial payment on New/Renewal membership and  https://github.com/civicrm/civicrm-core/pull/10352 has the fix. Currently, on core, partial payment for membership workflow is not working properly. 

Whereas made the fix for pledge whereas now irrespective of contribution_id if pledge id is present then fetch pledge statuses itself.

File: CRM/Member/Form/Membership.php
Comment (monishdeb) [inline CRM/Member/Form/Membership.php]: @seamuslee001 I don't think there is any requirement of those statuses irrespective of whether the membership has ```$_ _onlinePendingContributionId``` or not.  So we should simply exclude those two statuses for membership payment. What you say? These statuses are later excluded [here](https://github.com/civicrm/civicrm-core/pull/10718/files#diff-99d052aa6a966c5a2cc89d05f057d22dR545)

File: CRM/Contribute/BAO/Contribution/Utils.php
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution/Utils.php]: Can we switch to CRM_Pledge_BAO_Pledge::buildOptions? - that's where all the focus is  - getting that function right

File: CRM/Contribute/BAO/Contribution/Utils.php
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution/Utils.php]: ditto we should use buildOptions wherever possible

File: CRM/Contribute/BAO/Contribution/Utils.php
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution/Utils.php]: Actually it looks like these also duplicate further up

**Comment by eileenmcnaughton** [comment]
Hi all, I added a couple of comments. I think this PR makes sense over all & when @jitendrapurohit & @seamuslee001 are happy I think it's good to merge

File: CRM/Contribute/Form/Contribution.php
Comment (jitendrapurohit) [inline CRM/Contribute/Form/Contribution.php]: Removed variable `$statusName` is still in use at the [end of this function](https://github.com/civicrm/civicrm-core/blob/master/CRM/Contribute/Form/Contribution.php#L898) which would be undefined after the patch.

Comment (monishdeb) [inline CRM/Contribute/Form/Contribution.php]: @jitendrapurohit fixed it. Can you please check now?

**Review by jitendrapurohit — APPROVED** [review]
Thanks @monishdeb. This is working fine now. Tested on participant, contribution, membership form and seems safe to be merged @eileenmcnaughton

**Comment by eileenmcnaughton** [comment]
Per comment above I looked at this & was generally happy & Ok to merge once @jitendrapurohit confirmed, which he has now done. Merging

**Pull request merged**
